pkg/services/ngalert/notifier/alertmanager.go

```diff
@@ -66,36 +69,78 @@ func (am *Alertmanager) Run(ctx context.Context) error {
 		return errors.Wrap(err, "unable to initialize the silencing component of alerting")
 	}
 
-	am.notificationLog, err = nflog.New(
-		nflog.WithRetention(time.Hour*24),                         //TODO: This is a setting.
-		nflog.WithSnapshot(filepath.Join("dir", "notifications")), //TODO: This should be a setting
-	)
-	if err != nil {
-		return errors.Wrap(err, "unable to initialize the notification log component of alerting")
+	return nil
+}
+
+func (am *Alertmanager) Run(ctx context.Context) error {
+	// Make sure dispatcher starts. We can tolerate future reload failures.
+	if err := am.ReloadConfigFromDatabase(); err != nil {
+		return err
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			am.stopDispatcher()
+			return nil
+		case <-time.After(1 * time.Minute):
+			if err := am.ReloadConfigFromDatabase(); err != nil {
+				am.logger.Error("failed to sync config from database", "error", err)
+			}
+		}
 	}
+}
 
-	{
-		// Now, let's put together our notification pipeline
-		receivers := buildIntegrationsMap()
-		routingStage := make(notify.RoutingStage, len(receivers))
+func (am *Alertmanager) stopDispatcher() {
+	if am.dispatcher != nil {
+		am.dispatcher.Stop()
+	}
+	am.dispatcherWG.Wait()
+}
 
-		silencingStage := notify.NewMuteStage(silence.NewSilencer(am.silences, am.marker, gokit_log.NewNopLogger()))
-		//TODO: We need to unify these receivers
-		for name := range receivers {
-			stage := createReceiverStage(name, receivers[name], waitFunc, am.notificationLog)
-			routingStage[name] = notify.MultiStage{silencingStage, stage}
-		}
+// ReloadConfigFromDatabase picks the latest config from database and restarts
+// the components with the new config.
+func (am *Alertmanager) ReloadConfigFromDatabase() error {
+	am.reloadConfigMtx.Lock()
+	defer am.reloadConfigMtx.Unlock()
 
-		am.alerts, err = NewAlertProvider(routingStage, am.marker, gokit_log.NewNopLogger())
-		if err != nil {
-			return errors.Wrap(err, "failed to initialize alerting storage component")
-		}
+	// TODO: check if config is same as before using hashes and skip reload in case they are same.
+	cfg1, cfg2, err := getConfigFromDatabase()
+	if err != nil {
+		return errors.Wrap(err, "get config from database")
+	}
+	return errors.Wrap(am.reloadFromConfig(cfg1, cfg2), "reload from config")
```

Comment:
```suggestion
	return errors.Wrap(am.ApplyConfig(cfg1), "reload from config")
```

Reply:
I take it this is because `errors.Wrap(nil, "")` returns `nil`, right? It's fine, but I find this a bit too clever...

Reply:
Yes it returns nil, and it is a general pattern that I have seen.